000X-R-In-tree-pluggable-components.md

```diff
@@ -0,0 +1,132 @@
+
+When a user defines a component (in a Component YAML) that uses an in-tree pluggable component, during the initialization step of the component `daprd` will spawn a process that hosts the component as a pluggable component, and connects to it via UNIX Domain Socket (UDS), like any other pluggable component.
+
+Unlike "regular" pluggable components, `daprd` is the parent process that spawns the process hosting the component. `daprd` is also responsible for restarting the process if it crashes and can dispose of it if not needed anymore (e.g. in a future where components are more "dynamic").
+
+### Binaries
+
+Currently, all components are statically-compiled in `daprd` by creating a file in the [`cmd/daprd/components`](https://github.com/dapr/dapr/tree/release-1.10/cmd/daprd/components) package. Components that continues to be statically-compiled will not change.
+
+We will create a new package, for example `pkg/intree_pluggable`, which acts as host for pluggable components. It uses the Dapr pluggable component SDK to create the gRPC server on a socket, and to communicate with the component's code.
+
+Inside a folder such as `cmd/components/`, we will then create a folder for each component that becomes available as in-tree pluggable. In there, there's a `main` package that contains a `main()` function, which uses `pkg/intree_pluggable` and a registration code to create a binary that manages the component.
+
+These packages are compiled independently and generate separate binaries, one per each in-tree pluggable component.
+
+#### Example
+
+As a fictious example, let's look at a state store component that uses the Ethereum blockchain to store state 🙃
+
+The component's code is defined in `dapr/components-contrib` as any other component, for example in the `state/ethereum` package (in components-contrib).
+
+We then create a file in `dapr/dapr` in `cmd/components/state_ethereum/main.go` with content such as:
+
+```go
+package main
+
+import (
+	"github.com/dapr/components-contrib/state/ethereum"
+	pluggable "github.com/dapr/dapr/pkg/intree_pluggable"
+)
+
+func main() {
+	// Create a new in-tree pluggable component for a state store
+	component := pluggable.NewStateStore(logger)
+
+	// Register the component - just like we do in cmd/daprd/components
+	component.Register(ethereum.NewEthereumStore)
+	// Start the component, this is a blocking call
+	component.Run()
+}
+```
+
+> The code above is to be considered "pseudo-code" for reference only. The acutal interfaces of the internal packages are implementation details outside of the scope of this proposal.
+
+The component is then compiled like any other Go binary, creating a statically-compiled binary `intree_pluggable`.
+
+### Packaging
+
+The last step is packaging. After building all binaries, **they are included in the same container image as `daprd`**, for example in the `/components/` path within the container.
```

Comment:
If in this way, the image size is still big.

I have an idea that when dapr init, download all biniaries at each machine, `daprd` mount that dir when run. End users still deploy as before.

Reply:
I don't think we have concerns with the size of the image. What is most concerning to us is the size of the binary when is loaded in memory.

I am not a fan of downloading binaries on each machine, because:

1. It will make each container startup slower since the binaries need to be fetched over the network. If they're included in the Docker image, they are generally cached in the host already.
2. On K8s, it would require a volume claim, which is not ideal.
3. There are security and availability concerns in this case:
   - We will need to host a download server and make it always available
   - Users may have concerns with downloading code at runtime, as it could be used as an attack vector
   - Some users run Dapr on hermetic environments, where no network connections are allowed